File: frontend/src/types/loan.ts
// ── Loan Types ──────────────────────────────────────────────────

export type LoanStatus = 'active' | 'returned' | 'overdue';

export interface Loan {
  id: string;
  resource_id: string;
  book_copy_id: string | null;
  user_id: string;
  registered_by: string;
  loan_date: string;
  due_date: string;
  return_date: string | null;
  status: LoanStatus;
  notes: string | null;
}

export interface LoanWithUser extends Loan {
  user_name: string | null;
  user_email: string | null;
  user_career: string | null;
  user_group: string | null;
}

export interface LoanWithBook extends Loan {
  resource_title: string | null;
  resource_author: string | null;
  cover_url: string | null;
  copy_barcode: string | null;
}

export type LoanWithDetails = LoanWithUser & LoanWithBook;

export interface CreateLoanRequest {
  user_id: string;
  barcode: string;
  due_date: string;
  notes?: string | null;
}

export interface ReturnLoanRequest {
  notes?: string | null;
}

export interface BarcodeLookup {
  copy_id: string;
  barcode: string;
  status: string;
  resource_id: string;
  title: string;
  author: string | null;
  cover_url: string | null;
  isbn: string | null;
}

File: backend/app/schemas/loan.py
"""
Loan schemas.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class LoanResponse(BaseModel):
    """Loan details directly matching the DB schema."""
    id: str
    resource_id: str
    book_copy_id: Optional[str] = None
    user_id: str
    registered_by: str
    loan_date: datetime
    due_date: datetime
    return_date: Optional[datetime] = None
    status: str  # "active" | "returned" | "overdue"
    notes: Optional[str] = None


class LoanWithDetailsResponse(LoanResponse):
    """Loan response enriched with user and book details."""
    user_name: Optional[str] = None
    user_email: Optional[str] = None
    user_career: Optional[str] = None
    user_group: Optional[str] = None
    
    resource_title: Optional[str] = None
    resource_author: Optional[str] = None
    cover_url: Optional[str] = None
    
    copy_barcode: Optional[str] = None


class LoanCreateRequest(BaseModel):
    """Create a new loan."""
    user_id: str
    barcode: str  # Staff scans the barcode
    due_date: datetime
    notes: Optional[str] = None


class LoanReturnRequest(BaseModel):
    """Register a return."""
    notes: Optional[str] = None


class BarcodeLookupResponse(BaseModel):
    """Response when scanning a barcode before checkout."""
    copy_id: str
    barcode: str
    status: str  # "available" | "loaned" | "lost" ...
    resource_id: str
    title: str
    author: Optional[str] = None
    cover_url: Optional[str] = None
    isbn: Optional[str] = None
